File: backend/app/schemas/auth_schemas.py
"""
Authentication and user management validation schemas
Enhanced security with password strength validation
"""

from pydantic import BaseModel, EmailStr, Field, field_validator, ConfigDict
from typing import Optional
from datetime import datetime
import re


class UserRegister(BaseModel):
    """User registration with strict validation"""
    email: EmailStr = Field(
        ...,
        max_length=255,
        description="User email address"
    )
    password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        description="Password (min 8 chars, must include uppercase, lowercase, number)"
    )
    full_name: str = Field(
        ...,
        min_length=2,
        max_length=100,
        description="User's full name"
    )
    preferred_language: str = Field(
        default="es",
        pattern="^(es|en)$",
        description="Preferred language code (es or en)"
    )

    @field_validator('password')
    @classmethod
    def validate_password_strength(cls, v: str) -> str:
        """Validate password meets security requirements"""
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters long")

        if not re.search(r'[A-Z]', v):
            raise ValueError("Password must contain at least one uppercase letter")

        if not re.search(r'[a-z]', v):
            raise ValueError("Password must contain at least one lowercase letter")

        if not re.search(r'\d', v):
            raise ValueError("Password must contain at least one number")

        # Optional: Check for special characters
        if not re.search(r'[!@#$%^&*(),.?":{}|<>]', v):
            raise ValueError("Password must contain at least one special character")

        return v

    @field_validator('full_name')
    @classmethod
    def validate_full_name(cls, v: str) -> str:
        """Sanitize full name"""
        # Remove multiple spaces
        v = ' '.join(v.split())

        # Only allow letters, spaces, hyphens, and apostrophes
        if not re.match(r"^[a-zA-ZÀ-ÿ\s\-']+$", v):
            raise ValueError("Name contains invalid characters")

        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "email": "user@example.com",
                "password": "SecurePass123!",
                "full_name": "Juan García",
                "preferred_language": "es"
            }
        }
    )


class UserLogin(BaseModel):
    """User login credentials"""
    email: EmailStr = Field(
        ...,
        max_length=255,
        description="User email address"
    )
    password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        description="User password"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "email": "user@example.com",
                "password": "SecurePass123!"
            }
        }
    )


class UserResponse(BaseModel):
    """User data response (excludes sensitive info)"""
    id: int
    email: str
    full_name: str
    preferred_language: str
    is_active: bool
    created_at: datetime
    last_login: Optional[datetime] = None

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "email": "user@example.com",
                "full_name": "Juan García",
                "preferred_language": "es",
                "is_active": True,
                "created_at": "2025-10-03T12:00:00Z",
                "last_login": "2025-10-03T13:00:00Z"
            }
        }
    )


class TokenResponse(BaseModel):
    """JWT token response"""
    access_token: str = Field(..., description="JWT access token")
    token_type: str = Field(default="bearer", description="Token type")
    expires_in: int = Field(..., description="Token expiration time in seconds")
    user: UserResponse = Field(..., description="User information")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
                "token_type": "bearer",
                "expires_in": 86400,
                "user": {
                    "id": 1,
                    "email": "user@example.com",
                    "full_name": "Juan García",
                    "preferred_language": "es",
                    "is_active": True,
                    "created_at": "2025-10-03T12:00:00Z"
                }
            }
        }
    )


class UserUpdate(BaseModel):
    """User profile update"""
    full_name: Optional[str] = Field(
        default=None,
        min_length=2,
        max_length=100,
        description="Updated full name"
    )
    preferred_language: Optional[str] = Field(
        default=None,
        pattern="^(es|en)$",
        description="Updated preferred language"
    )

    @field_validator('full_name')
    @classmethod
    def validate_full_name(cls, v: Optional[str]) -> Optional[str]:
        """Sanitize full name"""
        if v is None:
            return v

        # Remove multiple spaces
        v = ' '.join(v.split())

        # Only allow letters, spaces, hyphens, and apostrophes
        if not re.match(r"^[a-zA-ZÀ-ÿ\s\-']+$", v):
            raise ValueError("Name contains invalid characters")

        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "full_name": "Juan Carlos García",
                "preferred_language": "en"
            }
        }
    )


class PasswordReset(BaseModel):
    """Password reset request"""
    email: EmailStr = Field(
        ...,
        max_length=255,
        description="User email address"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {"email": "user@example.com"}
        }
    )


class PasswordUpdate(BaseModel):
    """Password update with reset token"""
    token: str = Field(
        ...,
        min_length=20,
        max_length=500,
        description="Password reset token"
    )
    new_password: str = Field(
        ...,
        min_length=8,
        max_length=128,
        description="New password"
    )

    @field_validator('new_password')
    @classmethod
    def validate_password_strength(cls, v: str) -> str:
        """Validate password meets security requirements"""
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters long")

        if not re.search(r'[A-Z]', v):
            raise ValueError("Password must contain at least one uppercase letter")

        if not re.search(r'[a-z]', v):
            raise ValueError("Password must contain at least one lowercase letter")

        if not re.search(r'\d', v):
            raise ValueError("Password must contain at least one number")

        if not re.search(r'[!@#$%^&*(),.?":{}|<>]', v):
            raise ValueError("Password must contain at least one special character")

        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
                "new_password": "NewSecurePass123!"
            }
        }
    )

File: frontend/src/lib/validations/auth-schemas.ts
/**
 * Authentication Validation Schemas
 * Zod schemas for form validation with password strength checking
 */

import { z } from 'zod'

// Password strength validation
const passwordStrengthRegex = {
  lowercase: /[a-z]/,
  uppercase: /[A-Z]/,
  number: /\d/,
  special: /[!@#$%^&*(),.?":{}|<>]/,
}

export const passwordSchema = z
  .string()
  .min(8, 'Password must be at least 8 characters')
  .max(128, 'Password must not exceed 128 characters')
  .refine((password) => passwordStrengthRegex.lowercase.test(password), {
    message: 'Password must contain at least one lowercase letter',
  })
  .refine((password) => passwordStrengthRegex.uppercase.test(password), {
    message: 'Password must contain at least one uppercase letter',
  })
  .refine((password) => passwordStrengthRegex.number.test(password), {
    message: 'Password must contain at least one number',
  })
  .refine((password) => passwordStrengthRegex.special.test(password), {
    message: 'Password must contain at least one special character',
  })

// Login schema
export const loginSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: z.string().min(1, 'Password is required'),
})

export type LoginData = z.infer<typeof loginSchema>

// Registration schema
export const registerSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: passwordSchema,
  full_name: z
    .string()
    .min(2, 'Name must be at least 2 characters')
    .max(100, 'Name must not exceed 100 characters')
    .regex(/^[a-zA-ZÀ-ÿ\s\-']+$/, 'Name contains invalid characters'),
  preferred_language: z.enum(['es', 'en']).default('es'),
})

export type RegisterData = z.infer<typeof registerSchema>

// Password reset request schema
export const passwordResetRequestSchema = z.object({
  email: z.string().email('Invalid email address'),
})

export type PasswordResetRequestData = z.infer<typeof passwordResetRequestSchema>

// Password reset confirm schema
export const passwordResetConfirmSchema = z
  .object({
    token: z.string().min(20, 'Invalid reset token'),
    new_password: passwordSchema,
    confirm_password: z.string().min(1, 'Please confirm your password'),
  })
  .refine((data) => data.new_password === data.confirm_password, {
    message: 'Passwords do not match',
    path: ['confirm_password'],
  })

export type PasswordResetConfirmData = z.infer<typeof passwordResetConfirmSchema>

// Profile update schema
export const profileUpdateSchema = z.object({
  full_name: z
    .string()
    .min(2, 'Name must be at least 2 characters')
    .max(100, 'Name must not exceed 100 characters')
    .regex(/^[a-zA-ZÀ-ÿ\s\-']+$/, 'Name contains invalid characters')
    .optional(),
  preferred_language: z.enum(['es', 'en']).optional(),
})

export type ProfileUpdateData = z.infer<typeof profileUpdateSchema>

// Password strength calculator
export interface PasswordStrength {
  score: number // 0-7
  label: string
  color: string
}

export function getPasswordStrength(password: string): PasswordStrength {
  let score = 0

  // Length score (0-2 points)
  if (password.length >= 8) score++
  if (password.length >= 12) score++

  // Character variety (0-4 points)
  if (passwordStrengthRegex.lowercase.test(password)) score++
  if (passwordStrengthRegex.uppercase.test(password)) score++
  if (passwordStrengthRegex.number.test(password)) score++
  if (passwordStrengthRegex.special.test(password)) score++

  // Additional complexity (0-1 point)
  if (password.length >= 16) score++

  // Determine label and color
  let label: string
  let color: string

  if (score <= 2) {
    label = 'Weak'
    color = 'bg-red-500'
  } else if (score <= 4) {
    label = 'Fair'
    color = 'bg-yellow-500'
  } else if (score <= 5) {
    label = 'Good'
    color = 'bg-blue-500'
  } else {
    label = 'Strong'
    color = 'bg-green-500'
  }

  return { score, label, color }
}
